### PR TITLE
[dbnode] [query] Add num series/bytes received and decode datapoint time to traces

### PR DESCRIPTION
**Pull request by robskillington**
**What this PR does / why we need it**:

Needed instrumentation.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/query/functions/temporal/base.go [hunk 1]
```diff
@@ -223,14 +229,31 @@ func batchProcess(
...
 	mu *sync.Mutex,
 ) error {
+	var (
+		start      = time.Now()
+		decodeTime time.Duration
+	)
+	_, sp, _ := xcontext.StartSampledTraceSpan(ctx, tracepoint.TemporalDecodeParallel)
+	defer func() {
+		if decodeTime == 0 {
+			return // Do not record this span if instrumentation is not turned on.
```
Comment (notbdu) [inline src/query/functions/temporal/base.go]: Is there anything housekeeping we need to do with the unfinished unused span here?

File: src/query/block/series.go [hunk 2]
```diff
@@ -52,24 +54,38 @@ func (s Series) Len() int {
 
 // UnconsolidatedSeries is the series with raw datapoints.
 type UnconsolidatedSeries struct {
-	datapoints ts.Datapoints
-	Meta       SeriesMeta
+	opts UnconsolidatedSeriesOptions
```
Comment (rallen090) [inline src/query/block/series.go]: This is a little misleading with the series data itself inside an "option" struct now right? Seems like it might be better to have just a new 
```
type UnconsolidatedSeriesStats struct {
    Enabled         bool	
    DecodeTime time.Duration
}
```
within the original `UnconsolidatedSeries` struct

File: src/x/context/context.go [hunk 3]
```diff
@@ -321,57 +324,62 @@ func (c *ctx) parentCtx() Context {
...
 	ctx, sp, _ := c.StartSampledTraceSpan(name)
 	return ctx, sp
 }
+
+// StartSampledTraceSpan starts a span that may or may not be sampled and will
+// return whether it was sampled or not.
+func StartSampledTraceSpan(ctx stdctx.Context, name string) (stdctx.Context, opentracing.Span, bool) {
+	sp, spCtx := xopentracing.StartSpanFromContext(ctx, name)
+	sampled := spanIsSampled(sp)
+	fmt.Println("SAMPLED", sampled)
```
Comment (rallen090) [inline src/x/context/context.go]: Remove

File: src/x/context/context.go [hunk 4]
```diff
@@ -321,57 +323,61 @@ func (c *ctx) parentCtx() Context {
...
+func StartSampledTraceSpan(ctx stdctx.Context, name string) (stdctx.Context, opentracing.Span, bool) {
+	sp, spCtx := xopentracing.StartSpanFromContext(ctx, name)
+	sampled := spanIsSampled(sp)
+	if !sampled {
+		return ctx, noopTracer.StartSpan(name), false
+	}
+	return spCtx, sp, true
+}
+
+// Until OpenTracing supports the `IsSampled()` method, we need to cast to a Jaeger span.
```
Comment (arnikola) [inline src/x/context/context.go]: nit: comment not relevant any more

File: src/x/context/context.go [hunk 5]
```diff
@@ -321,57 +323,61 @@ func (c *ctx) parentCtx() Context {
...
+	if ok && jaegerSpCtx.IsSampled() {
+		return true
+	}
+
+	lightstepSpCtx, ok := spanCtx.(lightstep.SpanContext)
+	if ok && lightstepSpCtx.TraceID != 0 {
+		return true
+	}
+
+	mockSpCtx, ok := spanCtx.(mocktracer.MockSpanContext)
```
Comment (arnikola) [inline src/x/context/context.go]: nit: a bit iffy on having mock stuff in production code.. can we have the mocktracer extended a jaeger span context or something instead?

File: src/dbnode/client/types.go [hunk 6]
```diff
@@ -94,6 +94,13 @@ type Session interface {
 	Close() error
 }
 
+// FetchResponseMetadata is metadata about a fetch response.
+type FetchResponseMetadata struct {
+	Exhaustive         bool
```
[the remark this answers is not in the document]
Comment (rallen090) [reply src/dbnode/client/types.go]: Added

File: src/query/functions/temporal/base.go [hunk 7]
```diff
@@ -223,14 +228,31 @@ func batchProcess(
...
 	builder block.Builder,
 	blockMeta blockMeta,
 	processor processor,
 	mu *sync.Mutex,
 ) error {
+	var (
+		start      = time.Now()
+		decodeTime time.Duration
+	)
+	_, sp, _ := xcontext.StartSampledTraceSpan(ctx, tracepoint.TemporalDecodeParallel)
```
[the remark this answers is not in the document]
Comment (rallen090) [reply src/query/functions/temporal/base.go]: So `decodeTime` will always be initially zero here, and there is not a great way to get the `instrumented` flag in this processor component. However, I think we can create the span at the end since there is a `startTime` option, and so this should achieve the same behavior you're suggesting
```
                 _, sp, _ := xcontext.StartSampledTraceSpan(ctx, ..., opentracing.StartTime(start))
		sp.FinishWithOptions(opentracing.FinishOptions{
			FinishTime: start.Add(decodeTime),
		})
```

File: src/query/functions/temporal/base.go (hunk 1 above)
Comment (rallen090) [reply src/query/functions/temporal/base.go]: Changed so we only create the span if we intend to finish it

File: src/query/ts/m3db/consolidators/step_consolidator.go [hunk 8]
```diff
@@ -108,6 +106,7 @@ func (c *StepLookbackConsolidator) ConsolidateAndMoveToNext() float64 {
 	}
 
 	val := c.unconsumed[0]
-	c.unconsumed = c.unconsumed[1:]
+	copy(c.unconsumed, c.unconsumed[1:])
+	c.unconsumed = c.unconsumed[:len(c.unconsumed)-1]
```
Comment (arnikola) [inline src/query/ts/m3db/consolidators/step_consolidator.go]: This is only part of the fix here; should either revert and do a separate PR for this or get the whole fix in here instead.

File: src/dbnode/client/fetch_tagged_results_accumulator.go [hunk 9]
```diff
@@ -308,13 +328,17 @@ func (accum *fetchTaggedResultAccumulator) AsEncodingSeriesIterators(
 	})
 
 	exhaustive := accum.exhaustive && count <= limit && !moreElems
-	return result, exhaustive, nil
+	return result, FetchResponseMetadata{
+		Exhaustive:         exhaustive,
+		Responses:          len(accum.fetchResponses),
+		EstimateTotalBytes: accum.calcTransport.size,
```
Comment (arnikola) [inline src/dbnode/client/fetch_tagged_results_accumulator.go]: A little wary that `calcTransport` can be nil here; may be useful to add in a sanity check here, I think you can do something like this (similar model to [generated protobuf code](https://sourcegraph.com/github.com/m3db/m3/-/blob/src/query/generated/proto/rpcpb/query.pb.go#L226:33)) so that you don't have to add the check everywhere you'd want to access this:

```golang
func (c *calcTransport) GetSize() int {
 if c == nil {
  return 0
 }
 return c.size
}
```

File: src/dbnode/client/fetch_tagged_results_accumulator.go [hunk 10]
```diff
@@ -331,13 +355,17 @@ func (accum *fetchTaggedResultAccumulator) AsTaggedIDsIterator(
 	})
 
 	exhaustive := accum.exhaustive && count <= limit && !moreElems
-	return iter, exhaustive, nil
+	return iter, FetchResponseMetadata{
+		Exhaustive:         exhaustive,
+		Responses:          len(accum.aggResponses),
+		EstimateTotalBytes: accum.calcTransport.size,
```
Comment (arnikola) [inline src/dbnode/client/fetch_tagged_results_accumulator.go]: same here re: nil

File: src/dbnode/client/fetch_tagged_results_accumulator.go [hunk 11]
```diff
@@ -438,7 +466,11 @@ func (accum *fetchTaggedResultAccumulator) AsAggregatedTagsIterator(
...
```
Comment (arnikola) [inline src/dbnode/client/fetch_tagged_results_accumulator.go]: same here re: nil

File: src/query/ts/m3db/encoded_series_iterator.go [hunk 12]
```diff
@@ -89,6 +92,14 @@ func (it *encodedSeriesIter) Next() bool {
 		it.datapoints = it.datapoints[:0]
 	}
 
+	var (
+		decodeTime  time.Duration
```
Comment (arnikola) [inline src/query/ts/m3db/encoded_series_iterator.go]: nit: `decodeDuration` seems more accurate here

File: src/dbnode/client/fetch_tagged_results_accumulator.go [hunk 13]
```diff
@@ -241,6 +259,8 @@ func (accum *fetchTaggedResultAccumulator) Reset(
 			accum.shardConsistencyResults[id].enqueued++
 		}
 	}
+
+	accum.calcTransport.reset()
```
Comment (arnikola) [inline src/dbnode/client/fetch_tagged_results_accumulator.go]: Potentially nil here too

File: src/dbnode/client/fetch_tagged_results_accumulator.go [hunk 14]
```diff
@@ -212,6 +229,7 @@ func (accum *fetchTaggedResultAccumulator) Clear() {
 	accum.startTime, accum.endTime = time.Time{}, time.Time{}
 	accum.topoMap = nil
 	accum.exhaustive = true
+	accum.calcTransport.reset()
```
Comment (arnikola) [inline src/dbnode/client/fetch_tagged_results_accumulator.go]: Potentially nil here too

File: src/x/context/context.go (hunk 5 above)
Comment (arnikola) [reply src/x/context/context.go]: guess this'll have to be a necessary evil for now

File: src/query/ts/m3db/types.go [hunk 15]
```diff
@@ -78,6 +78,10 @@ type Options interface {
 	SetIteratorBatchingFn(IteratorBatchingFn) Options
 	// IteratorBatchingFn returns the batching function for the converter.
 	IteratorBatchingFn() IteratorBatchingFn
+	// SetInstrumented marks if the encoding step should have instrumentation enabled.
+	SetInstrumented(bool) Options
+	// Instrumented returns if the encoding step should have instrumentation enabled.
+	Instrumented() bool
```
Comment (arnikola) [inline src/query/ts/m3db/types.go]: nit: this may be more useful if it had an `instrument.Options` on it, and the instrument options had a `TraceEnabled` method?

File: src/query/storage/m3/storage.go [hunk 16]
```diff
@@ -434,15 +451,31 @@ func (s *m3storage) CompleteTags(
 	for _, namespace := range namespaces {
 		namespace := namespace // Capture var
 		go func() {
-			defer wg.Done()
+			_, span, sampled := xcontext.StartSampledTraceSpan(ctx,
```
Comment (arnikola) [inline src/query/storage/m3/storage.go]: Just double checking that this is threadsafe

File: src/query/functions/temporal/base.go [hunk 17]
```diff
@@ -279,11 +308,28 @@ func buildBlockBatch(
...
 	seriesIter block.SeriesIter,
 	builder block.Builder,
 	m blockMeta,
 	p processor,
 ) error {
+	var (
+		start      = time.Now()
+		decodeTime time.Duration
+	)
+	_, sp, _ := xcontext.StartSampledTraceSpan(ctx, tracepoint.TemporalDecodeSingle)
```
Comment (arnikola) [inline src/query/functions/temporal/base.go]: Should probably move this after the `decodeTime == 0` check in the defer function, similarly to how `parallelProcess` does it

File: src/query/storage/m3/storage.go (hunk 16 above)
Comment (rallen090) [reply src/query/storage/m3/storage.go]: Yeah I believe it is because it is not mutating anything, but rather returning a new context that wraps the pass-in one.

File: src/query/ts/m3db/types.go (hunk 15 above)
Comment (rallen090) [reply src/query/ts/m3db/types.go]: Decided to hold off on this til we have reason to expand this beyond tracing
